### PR TITLE
[3] Device manager: First version

### PR DESCRIPTION
**Pull request by patrickelectric**
Needs:
#499 
#505 

This **PR should be minimized and merged in a new branch** for internal tests (bot master), there is no reason to share this changes with the stable release branch/bug-fix.

### CODE REVIEW

**Comment by jaxxzer** [comment]
@patrickelectric to help with the review, please do a rebase after we merge [1] and [2].

**Comment by patrickelectric** [comment]
@jaxxzer this is ready to review, do not merge after approved, I would like to create a branch only for ping360 development.

**Comment by jaxxzer** [comment]
I know it's the first version. It looks attractive!

A bug:
1. open application, plug device by usb, connect to device at 115200 baud
1. open connection menu, connect to simultion
1. open connection menu, a device @9600 baud appears for selection in addition to the existing device

**Comment by patrickelectric** [comment]
@jaxxzer I was able to replicate the problem, it appears to be with protocol detection class.
```
Ping detected on: "LinkConfiguration{Name: Detector serial link, LinkType: 2, Arguments: (ttyUSB0:9600)}"
```
Will work in a solution.

**Comment by patrickelectric** [comment]
9600 baudrate fixed, will update the PR soon.

**Comment by patrickelectric** [comment]
- Can you point the steps to replicate the layout issue ?
- Add connection is a direct connection not a connection configuration for device, this same behavior was discussed before in old version manual connection.

**Comment by jaxxzer** [comment]
> Can you point the steps to replicate the layout issue ?

Resize the window smaller.

**Comment by jaxxzer** [comment]
> Add connection is a direct connection not a connection configuration for device, this same behavior was discussed before in old version manual connection.

Yes, I know. What is this in response to?

**Comment by jaxxzer** [comment]
> When entering information for udp ip/port, you must press ENTER while the entry field has focus, for your text entry to take effect (this should not be needed). 

Peek 2019-05-24 13-55

**Comment by patrickelectric** [comment]
> Yes, I know. What is this in response to?

image

Tested with:
 - Width: 619
 - Height: 362

**Comment by jaxxzer** [comment]
> Tested with:
Width: 619
Height: 362

How do I test with these numbers?

**Comment by patrickelectric** [comment]
> How do I test with these numbers?

I just tested with an arbitrary size, did the window resize with my mouse. 
You can print the window size to test.

**Comment by jaxxzer** [comment]
> You can print the window size to test.

How?

**Comment by patrickelectric** [comment]
@jaxxzer you Ping-Viewer commit version is different from the one of this PR, please test with the last version available.

**Comment by patrickelectric** [comment]
`main.qml`
`print(width, height)`

**Comment by jaxxzer** [comment]
> @jaxxzer you Ping-Viewer commit version is different from the one of this PR, please test with the last version available.

I just pushed to your branch. I had to bugfix because I was unable to run the application here.

```
$ git log --oneline
5761ae3 networktool: bugfix index out of range
6644ec8 ping-protocol-cpp: Update submodule
```

**Comment by jaxxzer** [comment]
> Can you point the steps to replicate the layout issue ?

Peek 2019-05-24 14-11

**Comment by jaxxzer** [comment]
Do we get to EDIT, DISCONNECT, or DELETE link configurations?

> image

**Comment by patrickelectric** [comment]
> Do we get to EDIT, DISCONNECT, or DELETE link configurations?

Not now, this can be done in the future.

**Comment by patrickelectric** [comment]
Manual connection layout bind fixed:
image

**Comment by patrickelectric** [comment]
Manual connection style bind fixed:
out

**Comment by patrickelectric** [comment]
> Manually entered connections should be saved/loaded (right now they are gone after you close the app)

This can be done in future versions.

**Comment by 9600** [comment]
You called?

**Comment by patrickelectric** [comment]
@9600 your nickname is a common serial baud rate, sorry for tagging you here !
image

File: src/devicemanager/devicemanager.h [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+signals:
+    void countChanged();
+    void sensorChanged(int objIndex);
+    void primarySensorChanged();
+
+private:
+    Q_DISABLE_COPY(DeviceManager);
+
+    /**
+     * @brief Construct a new Style Manager object
```
Comment (jaxxzer) [inline src/devicemanager/devicemanager.h]:
```suggestion
     * @brief Construct a new Device Manager object
```

File: src/devicemanager/devicemanager.h [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+     * @param scriptEngine
+     * @return QObject*
+     */
+    static QObject* qmlSingletonRegister(QQmlEngine* engine, QJSEngine* scriptEngine);
+
+    /**
+     * @brief Enables protocol detector
+     *
+     */
+    Q_INVOKABLE void start();
```
Comment (jaxxzer) [inline src/devicemanager/devicemanager.h]:
```suggestion
    Q_INVOKABLE void startDetecting();
```

File: src/devicemanager/devicemanager.h [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+    /**
+     * @brief Disables protocol detector
+     *
+     */
+    Q_INVOKABLE void stop();
```
Comment (jaxxzer) [inline src/devicemanager/devicemanager.h]:
```suggestion
    Q_INVOKABLE void stopDetecting();
```

**Pull request merged**